Adding Fork name tags to each Fork branch

sahilTakiar opened the pull request:
@ibuenros is this the correct approach?

Also, I see that `METRIC_CONTEXT_NAME_KEY` is set for `JobMetrics` on line 692 of `MRJobLauncher`, but I don't see an equivalent setting for `TaskMetrics`. Is this expected?

ibuenros commented on gobblin-runtime/src/main/java/gobblin/runtime/Fork.java:
```diff
@@ -103,8 +107,10 @@
+      CountDownLatch countDownLatch) throws Exception {
+
+    super(taskState, Fork.class.getClass(), createMetricsTags(taskState, index));
+    Instrumented.setMetricContextName(taskState, this.getMetricContext().getName());
```
You need to register the metric context before doing this (JobMetrics and TaskMetrics do it by default, but not Instrumented). You can do `GobblinMetricsRegistry.getInstance().putIfAbsent(context.getName(), context)`.

sahilTakiar replied:
Addressed, see comments below.

ibuenros commented on gobblin-runtime/src/main/java/gobblin/runtime/Fork.java:
```diff
@@ -60,7 +64,7 @@
  */
 @SuppressWarnings("unchecked")
-public class Fork implements Closeable, Runnable, FinalState {
+public class Fork extends Instrumented implements Closeable, Runnable, FinalState {
```
Can you make sure the method close() calls super.close()?

sahilTakiar replied:
I hold a reference to `MetricContext` and close it in the `Fork.close()` method.

ibuenros commented on gobblin-runtime/src/main/java/gobblin/runtime/Fork.java:
```diff
@@ -103,8 +107,10 @@
+
+    super(taskState, Fork.class.getClass(), createMetricsTags(taskState, index));
+    Instrumented.setMetricContextName(taskState, this.getMetricContext().getName());
 
```
Also, this will mutate the original input TaskState, which might not be a good idea if it is used elsewhere. Can we make a copy of the state instead?

sahilTakiar replied:
Created a copy constructor in `TaskState`

ibuenros commented:
About the use of METRIC_CONTEXT_NAME_KEY. It seems that this constant is defined in two different places: in ConfigurationKeys, and also in Instrumented. That is a mistake, we should use only one of them.

sahilTakiar commented:
With the current code base, its a little weird to have a class extend `Instrumented` and to get it registered in `GobblinMetricsRegistry`. When one extends `Instrumented`, the super constructor invokes `getMetricContext` which creates a `MetricContext` properly. However, the `MetricContext` is not registered in `GobblinMetricsRegistry`. In order to get it registered, I tried to create a `GobblinMetrics` instance using the `GobblinMetrics.get()` method (which automatically registers the newly made `MetricContext` in `GobblinMetricsRegistry`). However the `get()` method requires a `parentContext` be specified. This is where the problem arises, the `Instrumented` constructor has already made a `metricContext` that is a child of `TaskMetrics`, and now `GobblinMetrics` requires a `parentContext` to be specified.

So instead I solely use `GobblinMetrics`.

sahilTakiar commented:
@ibuenros can you take another look?

ibuenros commented on gobblin-runtime/src/main/java/gobblin/runtime/Fork.java:
```diff
@@ -102,32 +107,33 @@
     this.taskContext = taskContext;
-    this.taskState = taskState;
-    this.taskId = taskState.getTaskId();
+    this.taskStateCopy = new TaskState(taskState);
```
This could just be `this.taskState = new TaskState(taskState)`, given that we are not also storing the original. But I don't really have a preference either way.

ibuenros commented:
It looks good. Have you tested that the metric contexts do get generated as expected?

sahilTakiar commented:
I tested it, it is working as expected. The only change I had to make was to remove the copying of `TaskState` it seems that the `Fork` class is already mutating the `TaskState` object in a very other places, and there are some features that depend on the ability of mutating `TaskState`. Basically, a bunch of tests failing when I added the copying of `TaskState` so I removed it.
